File: tezrok-api/src/main/java/io/tezrok/api/builder/BaseBuilder.java
package io.tezrok.api.builder;

import io.tezrok.api.GeneratorContext;
import io.tezrok.api.model.node.ModuleNode;
import io.tezrok.api.model.node.ProjectNode;
import org.apache.commons.lang3.Validate;
import org.apache.velocity.Template;
import org.apache.velocity.VelocityContext;

import java.io.Writer;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

public abstract class BaseBuilder implements Builder {
    private final GeneratorContext generatorContext;
    private final ModuleNode module;

    protected BaseBuilder(ModuleNode module, GeneratorContext generatorContext) {
        this.generatorContext = Validate.notNull(generatorContext, "generatorContext");
        this.module = Validate.notNull(module, "module");
    }

    @Override
    public void build(final Writer writer) {
        VelocityContext context = new VelocityContext();

        if (getContext().isGenerateTime()) {
            context.put("generateTime", LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME));
        }

        onBuild(context);

        getTemplate().merge(context, writer);
    }

    protected abstract void onBuild(VelocityContext context);

    protected abstract Template getTemplate();

    public void init() {
    }

    @Override
    public boolean isCustomCode() {
        return false;
    }

    public GeneratorContext getContext() {
        return generatorContext;
    }

    public ProjectNode getProject() {
        return getContext().getProject();
    }
}

File: tezrok-core/src/main/kotlin/io/tezrok/core/builder/BaseBuilder.kt
package io.tezrok.core.builder

import io.tezrok.api.ExecuteContext
import io.tezrok.api.builder.Builder
import org.apache.velocity.Template
import org.apache.velocity.VelocityContext
import java.io.Writer
import java.time.OffsetDateTime
import java.time.format.DateTimeFormatter

abstract class BaseBuilder(private val context: ExecuteContext) : Builder {
    override fun build(writer: Writer) {
        val velContext = VelocityContext()

        if (context.isGenerateTime()) {
            velContext.put("generateTime", OffsetDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME))
        }

        onBuild(velContext)

        getTemplate().merge(velContext, writer)
    }

    protected abstract fun onBuild(context: VelocityContext)

    protected abstract fun getTemplate(): Template

    override fun init() {
    }

    override fun isCustomCode(): Boolean = false
}
